Task Supervisor for message consumers

akoutmos opened the pull request:
This PR is an initial attempt at addressing issue #46. Still in the process of adding tests and solidifying the implementation, so leaving the PR in draft until complete (also looking for implementation feedback). 

## Description
Originally, the `GenRMQ.Consumer` module would call `spawn/1` to asynchronously execute a message handler. This change instead leverages `Task.Supervisor` to execute message processing handlers. This should be a transparent change to the end user as this does not impact how `GenRMQ.Consumer` concurrently executes message handlers.

This introduces  the usage of OTP 21's `handle_continue` to ensure that the `Task.Supervisor` is started before any mailbox messages are processed. As a result, gen_rmq would no longer be compatible with OTP releases prior to 21.

## Checklist
- [x] I have added unit tests to cover my changes.
- [x] I have improved the code quality of this repo. (refactoring, or reduced number of static analyser issues)
- [x] I have updated the documentation accordingly

mkorszun commented on .travis.yml:
```diff
@@ -3,9 +3,9 @@ elixir:
 
 otp_release:
-  - 20.0
   - 21.0
```
I am fine with dropping support for this version. We will have to release a new major version though.

spier replied:
Just curious, what are we winning by dropping support for this version?

mkorszun replied:
* we can use new features from otp21
* we can continue upgrading `amqp` library, which right now I believe requires otp21 as well

akoutmos replied:
The primary reason for the change was so that `handle_continue` could be used to ensure that the init parts are called before anything else in the GenServer.

spier replied:
Got it. Thanks!

spier replied:
@mkorszun feel free to resolve this conversation, unless you need it as a reminder to "release a new major version".

mkorszun commented on lib/consumer.ex:
```diff
@@ -298,6 +297,20 @@ defmodule GenRMQ.Consumer do
 
+  @doc false
+  @impl GenServer
+  def handle_info({:DOWN, _ref, :process, _pid, _reason}, state) do
```
Will we get here messages for tasks or supervisor?

akoutmos replied:
This should be coming from the task if it failed. I'm thinking about adding a telemetry event in there to capture the error so library users can track processing failures.

mkorszun commented:
@akoutmos looking good - left some questions.

akoutmos commented:
I added some preliminary tests for this functionality, updated the telemetry documentation, and also refactored the consumer a bit to be able to act upon Tasks that it has spawned under the supervisor. There are still a few more edge cases that I want to cover in the tests, so I will keep this in draft until I feel as though I have covered those code paths.

akoutmos commented:
I think this one is at a good point to review now. I added the ability for consumer processes to wait on in flight tasks under the Task.Supervisor before terminating and also added tests to validate the flow. I also added an additional telemetry event to report on when Tasks fail along with some tests to cover these paths.

One question I have is if we should expose an optional configuration for how long to wait on running Tasks. Currently I am just using the default value that is set for `yield_many` https://hexdocs.pm/elixir/Task.html#yield_many/2

vorce commented:
@akoutmos I think yes we have to let users chose the task timeout. I suspect users will have very different expectations on their processor, depending on their business logic.

akoutmos commented:
I'll add that tonight and leave the default as the yield_many value.

spier commented:
@akoutmos can I just say: You are awesome :) Great contributions in this PR (and elsewhere). Thanks so much!

akoutmos commented:
Thanks for the kind words @spier! I appreciate it :)

akoutmos commented:
Went ahead and added the `terminate_timeout` option for users to configure how long to wait on in-flight Tasks. I also updated the docs to let them know about the option.

mkorszun commented:
@akoutmos amazing job 🥇- I will reserve some time during the weekend to review it.

vorce commented on lib/consumer.ex:
```diff
@@ -276,26 +288,42 @@ defmodule GenRMQ.Consumer do
+        {:DOWN, ref, :process, _pid, reason},
+        %{module: module, config: config, running_tasks: running_tasks} = state
+      ) do
+    if Map.has_key?(running_tasks, ref) do
```
What happens in this case to the message? Will it be rejected/nacked?

akoutmos replied:
Good call....we would probably want some sort of configuration so the user can decide what happens to the message if the task fails. Perhaps we also provide the option to send the message to the deadletter exchange. Also on that same note, we may also want some configuration around how long the Tasks take to complete. Currently it is set to the `async_nolink` value of 5 seconds. Thoughts?

vorce replied:
Maybe you should be able to set a callback to deal with the message? But I think the sensible default here would be to nack the message.

mkorszun replied:
I like the callback, smth like: `handle_error(reason, state)`. Then in this callback user has all the power to ack, reject or requeue.

mkorszun replied:
btw, are we capturing here also task timeouts?

akoutmos replied:
My latest commit adds the ability to configure a timeout for tasks. It was a little tricky to add that in given that Task.Supervisor does not directly support it, but I think I have a good implementation in place to handle this functionality.

As for a custom callback to handle the failure, are you thinking about rewriting the consumer module to a macro so that we can leverage `defoverridable` if the user does not want to use the default error handler?

mkorszun replied:
@akoutmos I would prefer to avoid rewriting consumer module to a macro at this stage (just to limit the scope of changes).

Since these changes will be released as a major version, requiring error callback to be implemented should be fine, right? @vorce what do you think?

We could also consider skipping error callback and just reject the messages. The problem here is that users might skip dead-letter configuration for their consumers.

vorce replied:
Hm hm. Yeah I think making each consumer implement a callback is OK. At least then we encourage users to think about this case. We can have a good implementation in the example consumer that logs and nacks the message.

akoutmos replied:
Sounds good. I'll add an additional callback `handle_error` that will be triggered on task exception or task timeout.

vorce replied:
Awesome, and thanks again for all the work @akoutmos

vorce commented on test/gen_rmq_consumer_test.exs:
```diff
@@ -92,6 +97,80 @@ defmodule GenRMQ.ConsumerTest do
+    test "should invoke the consumer's handle_info callback if error exists",
+         %{consumer: consumer_pid, state: state} = context do
+      clear_mailbox()
+      message = %{"value" => 0}
```
Can we make it clearer how this triggers an error in the consumer? Even just a comment would help clarify that this will trigger a division by zero.
Or make it even more explicit in ErrorConsumer with something like `if message.value == 0, do: raise "Error!"`

mkorszun commented on lib/consumer.ex:
```diff
@@ -59,12 +59,16 @@ defmodule GenRMQ.Consumer do
   Should be set to a value from `1..255` range. If it is greater than `255`, queue
   max priority will be set to `255`. Values between `1` and `10` are
-  [recommened](https://www.rabbitmq.com/priority.html#resource-usage).
+  [recommended](https://www.rabbitmq.com/priority.html#resource-usage).
```
👍

mkorszun commented on README.md:
```diff
@@ -75,7 +75,7 @@ This will result in:
 - durable topic `gen_rmq_exchange` exchange created or redeclared
 - durable `gen_rmq_in_queue` queue created or redeclared. It will be bound to `gen_rmq_exchange` exchange and has a deadletter exchange set to `gen_rmq_exchange.deadletter`
-- every `handle_message` callback will executed in separate process. This can be disabled by setting `concurrency: false` in `init` callback
+- every `handle_message` callback will executed in separate supervised Task. This can be disabled by setting `concurrency: false` in `init` callback
```
`will executed in separate supervised Task` -> `will be executed in separate supervised Task`

akoutmos commented:
Another thing that I thought about today was that there is no way to throttle the consumer. With how it currently is, it can keep on consuming messages and spawning new tasks under the supervisor. Thoughts on adding an additional configuration to have the user set the number of concurrent tasks running? `max_concurrent_tasks` ?

vorce commented:
@akoutmos good point! I think we can introduce that later if it turns out to be a problem for some users. The current/master implementation has the same problem.

akoutmos commented:
I added the `handle_error` callback for consumers that are configured to execute via Tasks. I'll need to also hook it up to trigger the `handle_error` callback when running in synchronous mode. But wanted to get some feedback on the approach before adding that in as well.

mkorszun commented on lib/consumer.ex:
```diff
@@ -186,6 +187,28 @@ defmodule GenRMQ.Consumer do
+  `reason` - atom denoting the type of error
+
+  ## Examples:
+  To reject the message message that caused the Task to fail you can do something like so:
```
double `message`

mkorszun commented on lib/message_task.ex:
```diff
@@ -0,0 +1,25 @@
+defmodule GenRMQ.MessageTask do
```
Good idea to wrap it like this 👍

mkorszun commented:
@akoutmos error handling looks good 👍

akoutmos commented:
I wrapped up that last of the tests and added `handle_error` callback support for non concurrent consumers. I'll give this a final look over tomorrow...but I think from a functionality perspective there is nothing else on my todo list. Let me know if you want to see any more tests/features :)

mkorszun commented:
@akoutmos great, thank you 🙏 I plan to merge it today/tomorrow.

akoutmos commented:
Updated some docs and some minor error handling tweaks. Nothing else pops out at me for clean up. All good on this side

spier commented:
🎉

spier commented:
@mkorszun just curious, what is the release process for the library? I remember you mentioning that we would have to release a new major(?) version for this?

mkorszun commented:
@spier we need a major release because we have dropped support for some older versions of the OTP. Release process will involve:
* bumping lib version to 3.0.0
* documentation differences / migration steps between v2 and v3
* merging all above changes to master and do a github release
* publish to hex.pm

spier commented:
@mkorszun Thanks for the writeup.

I am going through the steps, to understand them better. I might write a new issue to document the process for ourselves, and additional admins in the future. I will try to review what you do for this 3.0.0 release and then try myself on writing this up if that is ok?

## Notes

- bumping lib version to 3.0.0 => at least in [mix.exs](https://github.com/meltwater/gen_rmq/blob/master/mix.exs) and `README.md`.
- documenting differences / migration steps between v2 and v3 => probably in the [wiki](https://github.com/meltwater/gen_rmq/wiki/Migrations)? I looked for migrations steps from v1 to v2 as a sample. maybe we did not write those then?
- merging all above changes to master and doing a github release => [releases](https://github.com/meltwater/gen_rmq/releases)
- publishing to hex.pm => https://hex.pm/packages/gen_rmq

## (unrelated) 30k party

Btw I realized that we are at `28173` **downloads all time** and currently at `803` **downloads last 7 days**. So we should hit the 30k downloads milestone in about 2.5 weeks. How does one throw a virtual party? :)

mkorszun commented:
@spier sounds good. Your notes are already quite precise :)

regarding `30k party` - I believe we need it 🎉
